mava/systems/gpo/anakin/rec_magpo.py

```diff
@@ -0,0 +1,804 @@
+            last_done = done.repeat(env.num_agents).reshape(config.arch.num_envs, -1)
+            done = jnp.expand_dims(done, (1, 2, 3, 4))
+            sable_hstates = tree.map(lambda hs: jnp.where(done, jnp.zeros_like(hs), hs), sable_hstates)
+
+            prev_done = last_timestep.last().repeat(env.num_agents).reshape(num_envs, -1)
+            transition = Transition(
+                prev_done, action, value, timestep.reward, log_prob, last_timestep.observation, last_hstates
+            )
+            hstates = HiddenStates_all(sable_hstates, last_hstates.policy_hidden_state)
+            learner_state = LearnerState(params, opt_states, key, env_state, timestep, last_done, hstates)
+            metrics = timestep.extras["episode_metrics"] | timestep.extras["env_metrics"]
+            return learner_state, (transition, metrics)
+
+        # Copy old hidden states: to be used in the training loop
+        prev_sable_hstates = tree.map(lambda x: jnp.copy(x), learner_state.hstates.sable_hidden_state)
+
+        # Step environment for rollout length
+        learner_state, (traj_batch, episode_metrics) = jax.lax.scan(
+            _env_step, learner_state, length=config.system.rollout_length
+        )
+        # Calculate advantage
+        params, opt_states, key, env_state, last_timestep, last_done, updated_hstates = learner_state
+        key, last_val_key = jax.random.split(key)
+        _, _, last_val, _ = sable_action_select_fn(  # type: ignore
+            params.guider_params, last_timestep.observation, updated_hstates.sable_hidden_state, last_val_key
+        )
+
+        def _calculate_gae(
+            traj_batch: Transition,
+            current_val: chex.Array,
+            current_done: chex.Array,
+        ) -> Tuple[chex.Array, chex.Array]:
+            """Calculate the GAE."""
+
+            def _get_advantages(
+                carry: Tuple[chex.Array, chex.Array, chex.Array], transition: Transition
+            ) -> Tuple[Tuple[chex.Array, chex.Array, chex.Array], chex.Array]:
+                """Calculate the GAE for a single transition."""
+                gae, next_value, next_done = carry
+                done, value, reward = (
+                    transition.done,
+                    transition.value,
+                    transition.reward,
+                )
+                gamma = config.system.gamma
+                delta = reward + gamma * next_value * (1 - next_done) - value
+                gae = delta + gamma * config.system.gae_lambda * (1 - next_done) * gae
+                return (gae, value, done), gae
+
+            _, advantages = jax.lax.scan(
+                _get_advantages,
+                (jnp.zeros_like(current_val), current_val, current_done),
+                traj_batch,
+                reverse=True,
+                unroll=16,
+            )
+            return advantages, advantages + traj_batch.value
+
+        advantages, targets = _calculate_gae(traj_batch, last_val, last_done)
```

Comment:
Could you use our GAE util function :pray: 
Once you do just double check that performance is still as expected as it's quite easy to mess up the alignment between dones/obs with it. See [here](https://github.com/instadeepai/Mava/blob/develop/mava/systems/ppo/anakin/ff_ippo.py#L95-L97) for an example, but it looks to me like it is aligned correctly already